### PR TITLE
Introduce Spark Batch Job debug feature into Intellij

### PR DESCRIPTION
**Pull request by wezhang**
Now, The Spark batch job debugging is supported as a Intellij
run configuration.

Signed-off-by: Zhang Wei <zhwe@microsoft.com>

### CODE REVIEW

**Comment by vscjavaci** [comment]
Can one of the admins verify this patch?

**Comment by wezhang** [comment]
Hit #421

**Comment by vscjavaci** [comment]
Can one of the admins verify this patch?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java [hunk 1]
```diff
@@ -271,6 +324,49 @@ private void tryToCreateBatchSparkJob(@NotNull final IClusterDetail selectedClus
         }
     }
 
+    public SparkBatchRemoteDebugJob tryToCreateBatchSparkDebugJob(@NotNull final IClusterDetail selectedClusterDetail) throws HDIException,IOException {
+        SparkBatchSubmission.getInstance().setCredentialsProvider(selectedClusterDetail.getHttpUserName(), selectedClusterDetail.getHttpPassword());
+
+        try {
+            SparkBatchRemoteDebugJob debugJob = SparkBatchRemoteDebugJob.factory(SparkSubmitHelper.getLivyConnectionURL(selectedClusterDetail), submissionParameter, SparkBatchSubmission.getInstance());
+
```
Comment (voidfunction) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java]: telemetry point here?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+
+    @NotNull
+    @Override
+    public String getRunnerId() {
+        return "SparkBatchJobDebug";
+    }
+
+    @Override
+    public GenericDebuggerRunnerSettings createConfigurationData(ConfigurationInfoProvider settingsProvider) {
+        return null;
```
Comment (voidfunction) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java]: >null [](start = 15, length = 4)

null here?

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobSubmissionState.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import com.microsoft.azure.hdinsight.spark.common.SparkSubmitAdvancedConfigModel;
+import com.microsoft.azure.hdinsight.spark.common.SparkSubmitModel;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.ArrayList;
+
+/**
+ * Created by zhwe on 5/18/2017.
+ */
```
Comment (voidfunction) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobSubmissionState.java]: author information should be removed.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java [hunk 4]
```diff
@@ -117,7 +120,7 @@ public SparkSubmissionParameter getSubmissionParameter() {
     }
 
     public void setSubmissionParameters(SparkSubmissionParameter submissionParameters){
-        this.submissionParameter = submissionParameter;
+        this.submissionParameter = submissionParameters;
```
Comment (ziyel) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java]: This line is fix in merged pr https://github.com/Microsoft/azure-tools-for-java/pull/481.
It has conflicts now.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/RemoteDebugRunConfiguration.java [hunk 5]
```diff
@@ -22,17 +22,107 @@
...
+                    .ifPresent(attribute -> parameter.setArtifactName(attribute.getValue()));
+            Optional.ofNullable(element.getAttribute(SUBMISSION_ATTRIBUTE_CLASSNAME))
+                    .ifPresent(attribute -> parameter.setClassName(attribute.getValue()));
+
+            model.setSubmissionParameters(parameter);
+        }));
+    }
+
+    @Override
+    public void writeExternal(Element rootElement) throws WriteExternalException {
```
Comment (voidfunction) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/RemoteDebugRunConfiguration.java]: >public void writeExterna [](start = 4, length = 24)

I suppose you're willing to use interface PersistentStateComponent. [http://www.jetbrains.org/intellij/sdk/docs/basics/persisting_state_of_components.html](http://www.jetbrains.org/intellij/sdk/docs/basics/persisting_state_of_components.html)

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java (hunk 1 above)
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java]: #490 created

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java (hunk 2 above)
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobDebuggerRunner.java]: Exactly. No issue here.

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/RemoteDebugRunConfiguration.java (hunk 5 above)
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/configuration/RemoteDebugRunConfiguration.java]: Not for here. Here is Run configuration, what should refer to http://www.jetbrains.org/intellij/sdk/docs/basics/run_configurations/run_configuration_management.html#persistence

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobSubmissionState.java (hunk 3 above)
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/run/SparkBatchJobSubmissionState.java]: Fixed!

File: PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java (hunk 4 above)
Comment (wezhang) [inline PluginsAndFeatures/azure-toolkit-for-intellij/src/com/microsoft/azure/hdinsight/spark/common/SparkSubmitModel.java]: Fixed

**Pull request merged**